### ISSUE TITLE
forking a source into multiple throughs then combining into a single sink

### ISSUE THREAD

**Issue by phated** [pull-stream/pull-stream#72]
I've been trying to solve this problem but I think I lack some knowledge of pull-streams (still a newbie).

I have a single source and a single sink (I don't actually control these), my goal is to feed the source into multiple through streams and then combine the results of each of them and pass into the sink. The order doesn't matter.

``` md
       / through A \
source - through B - sink
       \ through C /
```

``` js
// The code would look something like
pull(
  pull.values([1, 2, 3]),
  fork([
    pull.map(function(val) {
      return val * 2;
    }),
    pull.map(function(val) {
      return val * 3;
    }),
    pull.map(function(val) {
      return val * 4;
    })
  ]),
  // I'd be fine with pull.flatten() here if that's the only way to make it work
  pull.collect(function(err, values) {
    // values should contain 2, 4, 6, 3, 6, 9, 4, 8, 12 in any order
  })
)
```

My current solution is:

``` js
'use strict';

var pull = require('pull-stream');
var Notify = require('pull-notify');
var through = require('pull-through');

function fork(throughs) {
  var notify = Notify();

  if (!Array.isArray(throughs)) {
    throughs = [throughs];
  }

  throughs.forEach(function(stream) {
    var queue;

    pull(
      notify.listen(),
      pull.map(function(event) {
        queue = event.queue;
        return event.data;
      }),
      stream,
      pull.drain(function(data) {
        if (queue) {
          queue(data);
        }
      }, function() {
        queue = null;
      })
    );
  });

  return through(function(data) {
    notify({ data: data, queue: this.queue });
  }, function() {
    notify.end();
    this.queue(null);
  });
}

module.exports = fork;
```

However, I'm still testing it in my application.

The important thing to note is that I can't use `pull-tee` or `pull-many` because I am not in control of my source or sink.

**Comment by dominictarr** [comment]
so, basically you want to create a take an array of transform pull-streams, and return a transform?
I don't really understand why you want the data to go through parallel throughs? Also, if you are using pull-notify in that way, it means the same chunks go through each stream.

okay I think you could do this using pull-tee, pull-pair and pull-many

``` js
var Pair = require('pull-pair')
var Many = require('pull-many')
var Tee = require('pull-tee')
function parathrough (throughs) { 
   var many = Many()
  return function (source) {
    pull(source, Tee(throughs.map(function (th) {
      var p = Pair()
      many.add(pull(p.source, th))
      return p.sink
    }))
    return many
})
}
```
the key is that a _through_ is just a _sink_ that returns a _sink_

**Comment by phated** [comment]
@dominictarr cool.  My actual usage (not the `pull.map` code above) looks something like:

```js
pull(
  // coming from a source out of my control
  pull.values([
    { type: 'add', data: {} },
    { type: 'remove', data: {} }
  ]),
  // within my control
  fork([
    handleAdd(),
    handleRemove(),
    handleBoth()
  ]),
  // out of my control
  pull.drain(...)
)

function handleAdd() {
  return pull(
    pull.filter(function(evt) {
      return evt.type === 'add';
    }),
    pull.map(function(evt) {
      // handle add events in some way
    })
  );
}

function handleRemove() {
  return pull(
    pull.filter(function(evt) {
      return evt.type === 'remove';
    }),
    pull.map(function(evt) {
      // handle remove events in some way
    })
  );
}

function handleBoth() {
  return pull(
    pull.filter(function(evt) {
      return evt.type === 'add' || evt.type === 'remove';
    }),
    pull.map(function(evt) {
      // handle add and remove events in some way
    })
  );
}
```

**Comment by dominictarr** [comment]
Can you describe what the app does? that would make it easier to understand why you are wanting to have the streams in parallel like this. Since it looks like you are doing sync transforms anyway, I think you could do it all simpler by say, using series of streams that was say, you could call it a `filteredMap`

it would take a filter and a map function, and if the filter returned true, it applies the map, otherwise it just passes through. That would end up having much simpler code than forking the stream only to recombine it.

**Comment by phated** [comment]
@dominictarr oh, it's definitely an over-engineered solution to a problem that doesn't really exist. It could (and probably should) be solved in your suggested way.  It's just a little toy project I'm working on and I wanted to push https://github.com/nrn/redux-pull-actors to it's limit.

I got a pretty clear understanding of how choo/inu would work by writing some elmlang but I wasn't sure how the actor pattern would hold up on top of redux.

Maybe @nrn has another example for usage of this pattern, as he mentioned it was something he expected of pull-streams.

**Comment by phated** [comment]
Just gave the code from https://github.com/pull-stream/pull-stream/issues/72#issuecomment-252120958 a try and it didn't work; however, my current solution produces the correct outcome.

**Comment by nichoth** [comment]
Splitting a stream seems useful if you have a source that is heavily multiplexed, like say a websocket that emits all events for your application. That way you could split it into different streams that go to different domains. I started experimenting with this in pull-mux, but not sure how it is going.
